jwst/outlier_detection/imaging.py

```diff
@@ -82,46 +88,48 @@ def detect_outliers(
     else:
         # for non-dithered data, the resampled image is just the original image
         drizzled_models = input_models
-        for i in range(len(input_models)):
-            drizzled_models[i].wht = build_driz_weight(
-                input_models[i],
-                weight_type=weight_type,
-                good_bits=good_bits)
-        # copy for when saving median and input is a filename?
-        median_wcs = copy.deepcopy(input_models[0].meta.wcs)
+        with input_models:
+            for i, model in enumerate(input_models):
+                model.wht = build_driz_weight(
+                    model,
+                    weight_type=weight_type,
+                    good_bits=good_bits)
+                # copy for when saving median and input is a filename?
+                if i == 0:
+                    median_wcs = copy.deepcopy(model.meta.wcs)
+                input_models.shelve(model, modify=True)
 
     # Perform median combination on set of drizzled mosaics
-    median_data = create_median(drizzled_models, maskpt)
+    median_data = create_median(drizzled_models, maskpt, on_disk=not in_memory)
 
     if save_intermediate_results:
         # make a median model
-        with datamodels.open(drizzled_models[0]) as dm0:
+        with drizzled_models:
+            example_model = drizzled_models.borrow(0)
+            drizzled_models.shelve(example_model, modify=False)
+            #with datamodels.open(example_model) as dm0:
             median_model = datamodels.ImageModel(median_data)
-            median_model.update(dm0)
+            median_model.update(example_model)
             median_model.meta.wcs = median_wcs
+            del example_model
 
         save_median(median_model, make_output_path, asn_id)
         del median_model
     else:
         # since we're not saving intermediate results if the drizzled models
         # were written to disk, remove them
         if not in_memory:
-            for fn in drizzled_models._models:
-                remove_file(fn)
+            for fn in drizzled_models._members:
```

Comment:
I debated making `members` a public property for `ModelLibrary`. I think this usage here is a good example of where it would be useful.

I think it could be replaced with this suggestion (using only public API) but I'm not sure which is easier to understand and/or maintain.
```suggestion
            for fn in drizzled_models.asn["products"][0]["members"]:
```

Reply:
I've never understood why associations need `"products"` in the first place.  I've never seen an association with multiple products.  Maybe I just need to read the association manual more closely.

So for that reason, I find the `.members` property much easier to understand, and I find it very annoying to have to go through the entire tree when`["products"][0]` is _always_ what we want to access

Reply:
nevertheless I am changing this as I agree we should use public API

Reply:
I completely agree :)

I wouldn't be opposed to:
- adding a `members` property here in the jwst `ModelLibrary` that we can back-propagate to `stpipe`
- leaving this as `_members` until we add a `members` property in `stpipe`
- changing this to `products[0]['members']` and adding a property in `stpipe` and we can then update this line

In all cases I think we should add a `members` property in `stpipe`.

https://github.com/spacetelescope/stpipe/issues/183

Reply:
ok I already changed it to use the public API, but depending on how long this review goes on, I can change it to library.members once that is merged into stpipe